backend/backend.go

```diff
@@ -40,7 +42,7 @@ func NewBackend(cli *client.Client) Backend {
 
 func (b Backend) RunExporter(ctx context.Context, exporter models.Exporter) {
 	var err error
-	ctx = context.WithValue(ctx, "step", stepCreate)
+	ctx = context.WithValue(ctx, "step", stepPullImage)
```

Comment:
To be sure I get it, this is where the `step` value is defined ? `stepPullImage` being the default val ?
Is `ctx` the object holding the docker daemon signals ?

Reply:
> To be sure I get it, this is where the step value is defined ? stepPullImage being the default val ?

Yep, exactly!

> Is ctx the object holding the docker daemon signals ?

Contexts, in go, are used to describe the execution context of goroutines. It's first use case is the ability to cancel/timeout a set of go routines, like DB requests when a HTTP request times out (see https://blog.golang.org/context).

Since Docker client methods need a context (as it's doing HTTP requests), this `ctx` got passed along the whole stack of function calls. In that way, when a container to export die during the startup process of its exporter, the associated context got cancelled, thus the HTTP request to Docker is cancelled and the `runStartupProcess` stops prematurely (and only then, the exporter cleanup starts).

But, as you said, this is also used to store the current step of the startup process (and other transient values), since `ctx` is passed along the whole call stack. I'm not sure if that's really a good idea :thinking: but for now this is how values are passed from a startup step to another.

Reply:
Thank you for the clarification :+1: Indeed, it could be good to have two objects (or an object with two keys) to split the `ctx` of the container to export from the `stratupProcess` of its exporter. Maybe find some other names for these vars, or start by writing some explanation doc so the scope of such vars would be easier to identify (and so to implement) ?